rosette/rbl/rosette/produce.rbl

```diff
@@ -0,0 +1,172 @@
+;;; Copyright (c) 2018, RChain Cooperative
+;;; Author: Kent Shikama <kent@kentshikama.com>
+;;; This file is licensed under the Apache License, version 2.0.
+;;;
+;;; File containing the reflective method produce and its helper methods
+
+(defProc (fetch-channel-pattern-and-product chart channel
+                                            product-patterns channel-index channel-position)
+  (if (= channel-index channel-position)
+      [channel -1]
+      (let [[[data-list continuation-struct-list] (tbl-get chart channel)]]
+        (if (null? data-list)
+            []
+            (iterate loop [[[data & r] data-list] [data-index 0]]
+                     (if (match-pattern (tuple-safe-nth product-patterns channel-index) data)
+                         [channel data-index]
+                         (loop r (inc data-index))))))))
+
+(defProc (fetch-pattern-and-product chart channels product-patterns channel-position)
+  (map channels
+       (proc [i channel]
+         (fetch-channel-pattern-and-product chart channel product-patterns i channel-position))))
+
+;;; Roughly dual to extract-data-candidates.
+;;; However, it is more complicated as it has to gather candidates across multiple channels.
+;;;
+;;; Roughly equivalent to the following python-ish pseudocode.
```

Comment:
I find I can't follow the python pseudocode either. Please document this with some non-code comments.